### PR TITLE
zjobs: misc combine fixes

### PR DESCRIPTION
**Pull request by Avokadoen**
- Fix compile issue when calling combine (at least for 0.10.0-dev.4249+11dce7894)
- Add test for combine
- Simplify combine implementation

### CODE REVIEW

**Comment by michal-z** [comment]
Thanks for the PR. @garettbass could you please help with review?

File: libs/zjobs/src/zjobs.zig [hunk 1]
```diff
@@ -478,23 +478,9 @@ pub fn JobQueue(
...
-                while (i < in.len and o < out.len) {
-                    out[o] = in[i];
-                    i += 1;
-                    o += 1;
-                }
-
-                id = try self.schedule(id, job);
-            }
+            var job = CombinePrereqsJob.init(self);
+            std.mem.copy(JobId, &job.prereqs, prereqs);
```
Comment (garettbass) [inline libs/zjobs/src/zjobs.zig:482]: What happens when `prereqs` is longer than `&job.prereqs`?  This is why I have a loop, it specifically creates a number of `CombinePrereqsJobs` to accomodate however many `prereqs` are supplied to this function.

File: libs/zjobs/src/zjobs.zig [hunk 2]
```diff
@@ -577,8 +563,15 @@ pub fn JobQueue(
             const prereq_size = @sizeOf(JobId);
             const max_prereqs = (max_job_size - jobs_size) / prereq_size;
 
-            jobs: *Self = .{},
-            prereqs: [max_prereqs]JobId = [_]JobId{JobId.none} ** max_prereqs,
+            jobs: *Self,
+            prereqs: [max_prereqs]JobId,
```
Comment (garettbass) [inline libs/zjobs/src/zjobs.zig:567]: I don't see any reason to move the initialization of `prereqs` into `init()`.  Please explain.

File: libs/zjobs/src/zjobs.zig [hunk 3]
```diff
@@ -1150,3 +1142,57 @@ test "JobQueue throughput" {
...
+            step.* = PrereqJob{ .counter = counter };
+        }
+
+        var job_ids: [pre_req_count]JobId = undefined;
+        for (chain_data) |step, i| {
+            job_ids[i] = try jobs.schedule(JobId.none, step);
+        }
+
+        const final = FinalJob{ .counter = counter };
+        const combined_ids = try jobs.combine(&job_ids);
```
Comment (garettbass) [inline libs/zjobs/src/zjobs.zig:1189]: Needs a test case where `job_ids.len` is greater than `CombinePrereqsJob.max_prereqs`.

File: libs/zjobs/src/zjobs.zig [hunk 4]
```diff
@@ -577,8 +563,15 @@ pub fn JobQueue(
             const prereq_size = @sizeOf(JobId);
             const max_prereqs = (max_job_size - jobs_size) / prereq_size;
 
-            jobs: *Self = .{},
```
Comment (garettbass) [inline libs/zjobs/src/zjobs.zig:580]: I'm surprised this would even compile?  Probably needed to be `undefined`?

**Review by garettbass — CHANGES_REQUESTED** [review]
1. Do not remove loop from `combine()` which creates enough `CombinePrereqsJobs` to store all of the provided `prereqs`, even if there are more than `CombinePrereqsJob.max_prereqs`.
2. Do not move initialization of `CombinePrereqsJob.prereqs` into `init()`, (or perhaps convince me this is the right thing to do).
3. Add a test case that `combines` more `prereqs` than will fit in a single `CombinePrereqsJob`, to ensure that case does not fail or panic.

**Comment by Avokadoen** [comment]
Thanks for looking at this. Ille fix the pr tomorrow after work. The init is there to fix a bug in xog that triggers a compile error on the original inline initialization

File: libs/zjobs/src/zjobs.zig (hunk 4 above)
Comment (Avokadoen) [reply libs/zjobs/src/zjobs.zig:580]: Can try this when i fix the other issues and try without init again 👍

File: libs/zjobs/src/zjobs.zig [hunk 5]
```diff
@@ -1150,3 +1142,57 @@ test "JobQueue throughput" {
     const throughput = @intToFloat(f64, job_ms) / @intToFloat(f64, main_ms);
     print("completed {} jobs ({}ms) in {}ms ({d:.1}x)\n", .{ job_count, job_ms, main_ms, throughput });
 }
+
+test "combine jobs respect prereq" {
```
Comment (garettbass) [inline libs/zjobs/src/zjobs.zig:1146]: Thanks for adding this, it was sorely needed!

**Comment by garettbass** [comment]
I merged your changes into my repo, and made the changes I had suggested.  I will create a new PR from my repo with your commits and my additional changes

**Comment by garettbass** [comment]
Superceded by **[PR-136](https://github.com/michal-z/zig-gamedev/pull/136)**

**Comment by michal-z** [comment]
> Superceded by **[PR-136](https://github.com/michal-z/zig-gamedev/pull/136)**

Should we close this one?

**Comment by Avokadoen** [comment]
Yes!

**Pull request closed without merging**